File: src/primitives/Hex/Hex.swift
import Foundation
import CVoltaire

/// Hex encoding/decoding utilities
public enum Hex {
    /// Convert hex string to bytes
    public static func decode(_ hex: String) throws -> [UInt8] {
        var buf = [UInt8](repeating: 0, count: hex.count / 2 + 1)
        let result = hex.withCString { ptr in
            primitives_hex_to_bytes(ptr, &buf, buf.count)
        }
        if result < 0 {
            throw VoltaireError(code: result)
        }
        return Array(buf.prefix(Int(result)))
    }

    /// Convert bytes to hex string (with 0x prefix)
    public static func encode(_ bytes: [UInt8]) -> String {
        var buf = [CChar](repeating: 0, count: bytes.count * 2 + 3)
        bytes.withUnsafeBufferPointer { ptr in
            _ = primitives_bytes_to_hex(ptr.baseAddress, ptr.count, &buf, buf.count)
        }
        return String(cString: buf)
    }

    /// Convert Data to hex string
    public static func encode(_ data: Data) -> String {
        encode([UInt8](data))
    }
}

File: src/primitives/Hex/Hex.kt
package com.voltaire

/** Hex encoding/decoding utilities */
object Hex {
    /** Decode hex string to bytes */
    @JvmStatic
    @Throws(VoltaireError::class)
    fun decode(hex: String): ByteArray {
        val maxLen = hex.length / 2 + 1
        val buf = ByteArray(maxLen)
        val result = VoltaireLib.INSTANCE.primitives_hex_to_bytes(hex, buf, maxLen.toLong())
        if (result < 0) throw VoltaireError.fromCode(result)
        return buf.copyOf(result)
    }

    /** Encode bytes to hex string with 0x prefix */
    @JvmStatic
    fun encode(data: ByteArray): String {
        val bufLen = data.size * 2 + 3
        val buf = ByteArray(bufLen)
        val result = VoltaireLib.INSTANCE.primitives_bytes_to_hex(data, data.size.toLong(), buf, bufLen.toLong())
        return if (result > 0) String(buf, 0, result) else "0x"
    }
}
